Add descriptions to vehicle structure

ClemensLinnhoff opened the pull request:
## Describe your changes
Add descriptions and coordinate frames to all elements in the vehicle structure


## Issue ticket number and link

Closes #102 

## Checklist before requesting a review

- [ ] I have performed a self-review of my code/documentation.
- [ ] My changes generate no new warnings during the documentation generation.

ClemensLinnhoff commented:
While adding the doors i realized, that the rotation axis is not defined. In the legacy the rotation axis was set to the z-axis and to the y-axis for the trunk lid. But there are also trunk lids opening around the z-axis.

We need to discuss this in one of the next meetings @LudwigFriedmann.

LudwigFriedmann commented:
Hi @ClemensLinnhoff, in legacy OpenMATERIAL, those axis were not aligned with x or z. Instead, the idea was to have virtual axis with arbitrary orientation that allow animation by single rotations.

ClemensLinnhoff commented:
Ahh I understand I will fix it.

ClemensLinnhoff commented:
@FKlopfer could you provide illustrations for the individual components?

I put notes like `NOTE: Add a figure ...` in multiple sections. I would like to have figures illustrating the coordinate system definitions specified the respective tables. You can have a look at the figures in the legacy OpenMAETERIAL for inspiration:
https://github.com/asam-ev/OpenMATERIAL/blob/OpenMATERIAL_Legacy/model_structure/Vehicle_Structure.md#33transforms-and-coordinate-frames

Please add the figures directly to this PR, then we have everything in the right place and it is easier to review because the figures are directly next to the specification.

I am not done with the definitions, yet. Currently I have everything until the lights. But I think you can already start with the images for the wheels and the doors.

FKlopfer commented:
@ClemensLinnhoff , we will have a look at it and come back to you if there are any questions.

ClemensLinnhoff commented:
@LudwigFriedmann, @MatthiasThDs, @ipg-sig, @KimuraDIVP, @DavidJRitter904, @MustafaTrian 

I have completed a first iteration of definition each individual element in the vehicle structure. Please have a look and review the content (naming, descriptions, coordinate definitions etc.).

Once we have agreed on the content, @FKlopfer can start an editorial review. Parson is also working on the figures in parallel.

MatthiasThDs commented:
I have 4 Questions:

1. Should we remove "GRP_" prefix, as its just everywhere the same?
2. Should we mention, that this is the Nodestructure, but users are allowed to attach multiple meshes (with arbitrary mesh-names) to each node?
3. I think we will define a small list of possible keywords for the parts? Because as a simulator, you may need to know what that part represents and how to connect your datastreams to those parts. 
4. Did you miss part_idx on Grp_Wheel_Steering_Rotating_<axle_idx>_<wheel_idx>?

ClemensLinnhoff commented:
1. I think you are right, we could remove the Grp_ prefix completely. What do the others think?
2. Yes, but we could put that in a general section, because this would also apply to other object classes.
3. I agree. I added a note, that we need to add a description and examples of keywords. I think we need to have a brief discussion about this. Because where to put the origin of the parts is also not clear, yet.
4. Good catch. I added it.

ipg-sig commented:
@ClemensLinnhoff Have we decided yet, that the wan´t to use the lowerCamelCase only for the metadata and stick with the snake_case "inside" the 3D geometry?
(I am in the process of getting up to date, but the topic was not adressed in the project level meeting, so I still have to look into the other meeting recordings...)

Regarding the questions:
1. Could be useful, if you want to set up the node structure automatically in your 3D application va script, so you can define that everything with the "GRP_" prefix should be an  group, empty,... Not sure if anyone would have the need for it though.
2. Agree with Clemens.
3. Yes, that would be useful and needed for the user. I think that should go into my Naming Convention issue, since those are predefined keys which should be named accordingly so it imakes life easier for everyone.

LudwigFriedmann commented:
To the points from @MatthiasThDs: 
1. I added the Grp_ prefix in the legacy OpenMATERIAL to clearly indicate group nodes (which are called empty nodes in Blender). The latter serve as transforms for their children and have to be identifiable by automated algorithms, so they should should have unique names.
The model_structure in the documentation shows only one half of the truth: In real-world examples, the structure is complemented with (child) nodes with arbitrary names. Therefore, I think that the Grp_ prefix still helps for human inspection to identify the relevant nodes right away.
2. Yes, we should definitely add that this node structure may have arbitrary numbers of arbitrary named nodes and meshes as children.
3. Shouldn't we used the node names from the structure right away for this purpose?
4. Looks legit

MatthiasThDs commented:
We could also think about **OMat_** instead of **Grp_**, when the idea is to identify Open Material Nodes. 
Thanks for clearing up, that users can even add more nodes if they want.

FKlopfer commented:
@ClemensLinnhoff , I have added the first three images. Could you please check if this is going in the right direction? I have taken the basic model of the vehicle from the legacy content.

ktrmpc commented:
1. I think it would be good to have a prefix that clearly identifies OpenMATERIALs. I think it would help distinguish related nodes.
2. I agree with making it clear that a node structure can have any number of arbitrarily named nodes and meshes as children.
3. I agree.

ipg-sig reviewed (commented):
I have discovered a few small things. If possible, please have a look before merging, but still can be adjusted later.

LudwigFriedmann commented on content/geometry/object-vehicle/vehicle-structure-model-structure.adoc:
```diff
@@ -5,68 +5,82 @@
 ----
 
-== Grp_Vehicle_Part_<keyword><vehicle_part_idx> (T) 
+== Grp_Vehicle_Part_<part_idx>_<keyword> (T) 
```
The (T) should indicate that there's a transform implemented in that node, i.e. that this node creates a local coordinate system.

ClemensLinnhoff replied:
Where should be put this explanation? This is true for all object classes.

ipg-sig replied:
I guess we should also add that to the general chapter, so that it´s clear for the reader.

ClemensLinnhoff replied:
On second thought: Do we really need this? The (T) is on almost every node. Alternatively we could just add a sentence to the ones that don't typically change in the simulation and say that they are fixed to the parent.

LudwigFriedmann commented on content/geometry/object-vehicle/vehicle-structure-model-structure.adoc:
```diff
@@ -5,68 +5,82 @@
-      |_ Grp_Rearview_Mirror_Mounting_<rearview_mirror_mounting_idx> (T)
-      |_ Grp_Rearview_Mirror_View_<rearview_mirror_view_idx> (T)
-      |_ Grp_Seat_<seat_idx> (T)
+Root
```
AFAIK this is still up for discussion.
In legacy openMATERIAL, the structure was 

Root
|-Exterior
|-Interior
|-Floor

So the root node had its purpose as a representation of the vehicle.

With the introducted of vehicle parts, if we deem the latter as self-contained, we could have them at the topmost hierarchy level.
Personally, I'd favor to have a "root" node as group node for the vehicle parts and as representation of the vehicle since I can't imagine using the vehicle parts stand-alone.

ClemensLinnhoff replied:
We decided that every object should have a root node. This explanation should go into a general chapter, as it is true for all object classes.

LudwigFriedmann commented on content/geometry/object-vehicle/vehicle-structure-model-structure.adoc:
```diff
@@ -5,68 +5,82 @@
+
+NOTE: Define the origin of the vehicle (part)
+
+NOTE: Add description and examples for keywords
```
Can you give examples for keywords? I hoped the node structure would be self-explanatory.

ipg-sig replied:
Examples I would imagine are something like Body, Cabin, Platform for cars and trucks. For motorcycles something like SFT (Suspension Front Top), SFB (Suspension Front Bottom), SRA (Suspension Rear Arm). 

But I would keep the keywords open to discussion for the next meeting, if we want to keep them open or not.

ipg-sig commented:
Maybe another point which just pops into my eye while i am preapring the naming convention part:
Would it be useful to change the order of the subkeys for the mirrors?
From "Grp_Side_Mirror_Mounting_Left" to "Grp_Mirror_Side_Mounting_Left"? With this approach we could be uniform with the lights, so we always have the component (light; mirrors) and after that the specification (brake or indicator; side or rear)?

FKlopfer commented on content/geometry/object-vehicle/vehicle-structure-model-structure.adoc:
```diff
@@ -162,27 +162,35 @@ Add a description
+Zero rotation and and translation around all axles are defined in neutral load conditions.
+In the 3D model, both caster and camber angles are neglected.
+
+image::Grp_Wheel.svg[,1100]
```
The image will only show up in the final antora generated documentation as the linking mechanism works differently for antora and the preview in Visual Studio Code. You can see the images in the "_images" folder or directly here on the PR.

ClemensLinnhoff replied:
This is true, but there is a solution. I added the image directory as a parameter, so your local IDE (tested with VS Code) can find and display the images. And also GitHub can find the images and display them directly (see [here](https://github.com/asam-ev/OpenMATERIAL/blob/102-add-description-to-vehicle-structure-elements/content/geometry/object-vehicle/vehicle-structure-model-structure.adoc#grp_wheel_part_idx_axle_idx_wheel_idx-t))

ClemensLinnhoff commented:
> @ClemensLinnhoff , I have added the first three images. Could you please check if this is going in the right direction? I have taken the basic model of the vehicle from the legacy content.

Thank you. Please make sure, that the coordinate frame fit to the description of the specific element. For example the coordinate frome on the front door (engine hood) does not comply to the description.

Please also make sure that all illustrations are consistent.
- The Light_Day_Right looks completely different for the one on the left.
- Also it would be nice, if the coordinate frame had the same size in all figures.
- Grp_Door_Top does not fit to the style of the other images at all.

ClemensLinnhoff commented on content/geometry/object-vehicle/vehicle-structure-model-structure.adoc:
```diff
@@ -5,68 +5,82 @@
 ----
 
-== Grp_Vehicle_Part_<keyword><vehicle_part_idx> (T) 
+== Grp_Vehicle_Part_<part_idx>_<keyword> (T) 
```
Fixed

ClemensLinnhoff commented on content/geometry/object-vehicle/vehicle-structure-model-structure.adoc:
```diff
@@ -5,68 +5,82 @@
+Therefore, only one Grp_Vehicle_Part exists with the name `Grp_Vehicle_Part_0`.
+
+The `<part_idx>` must be added to all sub-nodes, as multiple nodes with the same name in different vehicle parts are not possible.
+There cannot be a node `Grp_Exterior` in both `Grp_Vehicle_Part_0`` and `Grp_Vehicle_Part_1``.
```
Fixed

ipg-sig commented on content/geometry/object-vehicle/vehicle-structure-notes.adoc:
```diff
@@ -1,37 +1,7 @@
-*** Comment MT: MT: What about buses with middle axis? More complex geometries/functions? Multiple Trailers? movable Cabins?  
-*** Should add nodes for trailer coupling devices, so that you know where to attach (MT)  
\ No newline at end of file
+* Any need for moving pedals or shift levers, or the levers on the steering wheel? (SG)
```
No need at the moment from our side. Could be something for a later version and a more complex LOD level. but it is not necessary for the first version.
But I can add recommendations for subkeys for hitches and levers in the naming convention nonthless.

AsamDiegoSanchez reviewed (approved):
Reviewed and agreed.
Signed-off by: Diego Sánchez <diego.sanchez@asam.net>

ClemensLinnhoff commented:
Thanks for the reviews.
As @LudwigFriedmann proposed in the last meeting: let's merge this and do any more enhancements in separate branches.
